sdk/identity/azure-identity/inc/azure/identity/detail/client_credential_core.hpp

```diff
@@ -12,14 +12,14 @@
 #include <vector>
 
 namespace Azure { namespace Identity { namespace _detail {
+  constexpr auto AzureAuthorityHostEnvVarName = "AZURE_AUTHORITY_HOST";
+
```

Comment:
We could incapsulate all the defaults in a helper class, and the logic for handling the defaults won't be in the credential - options will have values, and user can create options and see the value it was initialized with in all cases:
```cpp
// User does not have AZURE_AUTHORITY_HOST env var set.
ClientSecretCredentialOptions options;
std::cout << options.AuthorityHost;
// ^^^^^^ would print  "https://login.microsoftonline.com/"
```
Credentials won't have to have logic to assign the default value to empty fields, there will be less lines to update.

```suggestion
class DefaultOptionValues final
{
  DefaultOptionValues() = delete;
  ~DefaultOptionValues() = delete;

public:
  static std::string GetAuthorityHost()
  {
    auto const envAuthHost = Core::_internal::Environment::GetVariable("AZURE_AUTHORITY_HOST");
    return envAuthHost.empty() ? "https://login.microsoftonline.com/" : envAuthHost;
  }
  
  // similar method can be added here for Tenand ID and so on:
  // static std::string GetTenantId()
}
```

And then, in the options:
```cpp
struct ClientSecretCredentialOptions ... {
  ...

  std::string AuthorityHost
    = _internal::DefaultOptionValues::GetAuthorityHost();
};
```

Reply:
This would be consistent with .NET: https://github.com/Azure/azure-sdk-for-net/blob/240a4fcab49587107c01bf60168357ba8be84b1e/sdk/identity/Azure.Identity/src/AzureAuthorityHosts.cs#L37-L45

Reply:
And Java: https://github.com/Azure/azure-sdk-for-java/blob/9c250fe98caea2920354357fa3fbaef373f6effc/sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java#L686-L687

Reply:
Will address separately https://github.com/Azure/azure-sdk-for-cpp/issues/5002